sda/overlay.go

```diff
@@ -123,12 +239,115 @@ func (o *Overlay) TransmitMsg(sdaMsg *ProtocolMsg) error {
 	return nil
 }
 
+// sendSDAData marshals the inner msg and then sends a Data msg
+// to the appropriate entity
+func (o *Overlay) sendSDAData(e *network.ServerIdentity, sdaMsg *ProtocolMsg) error {
+	b, err := network.MarshalRegisteredType(sdaMsg.Msg)
+	if err != nil {
+		typ := network.TypeFromData(sdaMsg.Msg)
+		rtype := reflect.TypeOf(sdaMsg.Msg)
+		var str string
+		if typ == network.ErrorType {
+			str = " Non registered Type !"
+		} else {
+			str = typ.String()
+		}
+		str += " (reflect= " + rtype.String()
+		return fmt.Errorf("Error marshaling  message: %s  ( msg = %+v)", err.Error(), sdaMsg.Msg)
+	}
+	sdaMsg.MsgSlice = b
+	sdaMsg.MsgType = network.TypeFromData(sdaMsg.Msg)
+	// put to nil so protobuf won't encode it and there won't be any error on the
+	// other side (because it doesn't know how to decode it)
+	sdaMsg.Msg = nil
+	log.Lvl4("Sending to", e.Addresses)
+	return o.host.SendRaw(e, sdaMsg)
+}
+
+// addPendingTreeMarshal adds a treeMarshal to the list.
+// This list is checked each time we receive a new Roster
+// so trees using this Roster can be constructed.
+func (o *Overlay) addPendingTreeMarshal(tm *TreeMarshal) {
+	o.pendingTreeLock.Lock()
+	var sl []*TreeMarshal
+	var ok bool
+	// initiate the slice before adding
+	if sl, ok = o.pendingTreeMarshal[tm.RosterID]; !ok {
+		sl = make([]*TreeMarshal, 0)
+	}
+	sl = append(sl, tm)
+	o.pendingTreeMarshal[tm.RosterID] = sl
+	o.pendingTreeLock.Unlock()
+}
+
+// checkPendingSda is called each time we receive a new tree if there are some SDA
+// messages using this tree. If there are, we can make an instance of a protocolinstance
+// and give it the message!.
+// NOTE: put that as a go routine so the rest of the processing messages are not
+// slowed down, if there are many pending sda message at once (i.e. start many new
+// protocols at same time)
+func (o *Overlay) checkPendingSDA(t *Tree) {
+	go func() {
+		o.pendingSDAsLock.Lock()
+		var newPending []*ProtocolMsg
+		for _, msg := range o.pendingSDAs {
+			// if this message references t
```

Comment:
If the comment describes the whole decision, put it outside. If the comment describes what happens after the decision, put it inside the if:
```
// check whether we received a message for tree t
if ...
```
or
```
if ...
    // received message for the tree t
else
    // not our message, append to pending messages
```